### PR TITLE
[4.3][Guided Tours] Move the guided tours system plugin functionality to the module and component

### PR DESCRIPTION
**Pull request by laoneo**
Alternative to #40037.

### Summary of Changes
This pr moves the functionality of the ajax request to the component and the asset loading to the module. Like that when the module is disabled there is no system plugin floating around. The first approach was to have a component which was for obvious reasons wrong.

~It removes the tours system plugin and puts the code to load tours to the template. I have the impression that the template must actively enable it, because it has some hardcoded bootstrap classes in it. The rest of the tours plugin code is moved to a controller.~

~@Fedik can you have a look on the changes in the index.php file, as I think there is room for improvements there. Guess the dependency can be declared on the templates ass.json file.~

### Testing Instructions
Walt through the tours.

### Actual result BEFORE applying this Pull Request
All works.

### Expected result AFTER applying this Pull Request
All works.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by brianteeman** [comment]
Unless I have misunderstood something doesnt this mean that it will only work with the atum template?

File: administrator/templates/atum/index.php [hunk 1]
```diff
@@ -62,6 +63,7 @@
 preg_match('#^hsla?\(([0-9]+)[\D]+([0-9]+)[\D]+([0-9]+)[\D]+([0-9](?:.\d+)?)?\)$#i', $this->params->get('hue', 'hsl(214, 63%, 20%)'), $matches);
 
 // Enable assets
+$wa->getRegistry()->addRegistryFile('media/com_guidedtours/joomla.asset.json');
```
Comment (Fedik) [inline administrator/templates/atum/index.php]: Change it to `...->addExtensionRegistryFile('com_guidedtours')`

Comment (laoneo) [reply administrator/templates/atum/index.php]: Can this somehow not be declared in the templates asset files dependencies?

**Comment by Fedik** [comment]
> Guess the dependency can be declared on the templates ass.json file.

It looks good. You loading another extension, what you made here is fine.

**Comment by laoneo** [comment]
> Unless I have misunderstood something doesnt this mean that it will only work with the atum template?

Yes exactly. As for me this feature has a tight relation to the template it has to actively enable it.

File: administrator/templates/atum/index.php (hunk 1 above)
Comment (Fedik) [reply administrator/templates/atum/index.php]: Nope, unfortunately, this what you should do manualy

File: administrator/templates/atum/index.php [hunk 2]
```diff
@@ -74,6 +74,8 @@
 		--template-special-color: ' . $this->params->get('special-color', '#001B4C') . ';
 	}');
 
+$app->bootComponent('com_guidedtours')->prepareAssets($app);
```
Comment (dgrammatiko) [inline administrator/templates/atum/index.php]: Some nitpicking:

```suggestion
if (ComponentHelper::isEnabled('com_guidedtours')) {
    $app->bootComponent('com_guidedtours')->prepareAssets($app);
}
```

Also needs an import at the top of the file: 
```php
use Joomla\CMS\Component\ComponentHelper;
```

**Review by dgrammatiko — CHANGES_REQUESTED** [review]

File: administrator/templates/atum/index.php (hunk 2 above)
Comment (laoneo) [reply administrator/templates/atum/index.php]: I would do it a bit different https://github.com/joomla/joomla-cms/pull/40046/commits/f58b1c8790b6f0cd1e994f5545f6f1a4aaf1d248

**Comment by brianteeman** [comment]
> > Unless I have misunderstood something doesnt this mean that it will only work with the atum template?
> 
> Yes exactly. As for me this feature has a tight relation to the template it has to actively enable it.

In that case this is a down vote from me as afaik there is no core functionality that is dependent on the template being used.

**Comment by bembelimen** [comment]
Ok, here is my personal opinion:
I agree, there is still stuff which can be improved in the Guided Tour and I really appreciate that you make your thoughts about it, but the criteria which are hold against the Guided Tour (and leading to this PR) are very hypocritical.

### No system plugin
Currently you can, if you don't want to have the tour active just disable the module + plugin. With this implementation you're stuck to boot the component always and also the assets are not disabled. Implementing it in the template always leads to a non-disable overhead.

Additional the PR from @obuisard fixes the problem of having stuff loaded in the frontend.

Looking into the plugin list that is the way we do it.

### Template binding
This is a move which kills the tour from the start, as it's first not documented and 2nd people don't understand it how to implement it. (Sorry I don't count "there are a few bootstrap classes, they're very minor as shepherd.js is bootstrap independed and I think it falls back to the button classes). If we put that meassurement as a criteria, we really have a problem with e.g. validation.

So this looks for me not the correct solution and it tries to fix things with solutions which just bring in other problems.

**Comment by laoneo** [comment]
As I was digging deeper into the code it became obvious that the asset loading must be in the module. So I changed this and the template is not touched anymore. Now it should be much more logical. @brianteeman can you have a look again, should be better now and more clearer.

**Comment by joomdonation** [comment]
At first, I thought using module is a smart solution. However, the problem is that when we are on create article screen for example, the module is not loaded, so the script.... are not loaded and it breaks the feature.

**Comment by laoneo** [comment]
Any idea how to solve that?

**Comment by joomdonation** [comment]
Afraid we would have to follow @wilsonge suggestion: Create a new plugin group and import it for administrator app only.

**Comment by laoneo** [comment]
It was hardcoded, that it doesn't show up in forms. Removed that restriction with https://github.com/joomla/joomla-cms/pull/40046/commits/fa89db6649dd05b1c848d9f283ee4e27927c49dc.

**Comment by brianteeman** [comment]
That restriction was by design but had consequences https://github.com/joomla/joomla-cms/issues/40025

**Comment by joomdonation** [comment]
Seem to work but still have some concerns:

- This assumes the module is assigned to a position which is always loaded in the template
- If you are entering data on the form and click on a tour, then the page is redirected and you lost the data entered in the form. Maybe that was the reason it is designed to not being loaded on forms.

**Comment by laoneo** [comment]
I can adapt the code that the assets are loaded but not the button. So we have the same behavior as with the system plugin. And yes, when the module is not loaded, then no guided tours :man_shrugging: .

**Comment by laoneo** [comment]
> That restriction was by design but had consequences #40025

With this approach we can enable or disable it and still make tours working correctly.

**Comment by Fedik** [comment]
I think correct approach it to have a component to handle all requests and all logic, and plugin to load asset. 
Because at current point of time we cannot have "single extension" that does both.

For me #40037 is okay (not perfect but okay), I would also move out all `startTour` logic to component controller.

**Comment by laoneo** [comment]
Why do you need a plugin which does both? I'm not getting it? The module starts the tour so it should be responsible for all of it.

**Comment by Fedik** [comment]
> Why do you need a plugin which does both?

I did not said that. I wrote about "single extension", eg. component that can listen to App events, kind of.
Well, nevermind, not really important :wink:

**Comment by Fedik** [comment]
Okay, maybe module is not that bad also.

**Comment by laoneo** [comment]
Makes indeed more sense to have the assets in the component folder. Moved it back.

**Comment by laoneo** [comment]
@richard67 can you help me with that?

**Comment by rdeutz** [comment]
Making a functionality depend on a module load doesn't seems right to me. It's not logical.

**Comment by richard67** [comment]
> @richard67 can you help me with that?

@laoneo DELETE FROM `#__extensions` WHERE `type` = 'plugin' AND `element` = 'guidedtours' AND `folder` = 'system' AND `client_id` = 0;

But maybe we should better uninstall it using the installer like it is done here with the EOS plugin: https://github.com/joomla/joomla-cms/blob/e462044fc9a6f795c67c02c98daff5fa7d71dcba/administrator/components/com_admin/script.php#L491 .

**Comment by laoneo** [comment]
For the current architecture this pr is absolutely ok as the module is initiating the whole process. Currently when you disable the module all the assets are still loaded which makes no sense at all. When more use cases will be implemented in the future (which I hope as I see a big potential in the guided tours feature) and having the functionality in a module is not anymore enough, then we can discus new approaches. But for now this is fine as it is.

**Comment by HLeithner** [comment]
With this module only approach it's not possible to use the guided tour in the frontend

**Comment by richard67** [comment]
@laoneo I see you have decided to use an update SQL script to delete the extension and not a new method in script.php for uninstalling it. To be honest I don’t know if that works. We have a system plugin which is deleted from the extensions table with the update SQL script and the files of that plugin will be removed with the deleted files thing in script.php. I don’t know from scratch now in which order of processing that happens. Worst case is we delete the plugin while it still is loaded. Please carefully test an update of a beta 4 where the guided tours are used to the update package of this PR. We should be 100% sure it doesn’t crash. I think there should be a reason why other plugins are uninstalled with a method in script.php instead of doing it with an update SQL. Maybe @wilsonge remembers why we did it that way for other plugins?

**Comment by laoneo** [comment]
> With this module only approach it's not possible to use the guided tour in the frontend

Making guided tours available on the front end is a much bigger task as there is no standard way to create the submit buttons etc. So to implement this it needs a much bigger refactoring. As I said, this change is for the current architecture and when we want to expand to the front end, it needs much more logic. Then it would be possible to think about plugins, for a custom group.

**Comment by joomdonation** [comment]
> wouldnt you just need a site module?

@brianteeman The problem is that the current implementation requires that the module is always loaded which is not always true (especially when it is used on frontend).

**Pull request closed without merging**